### PR TITLE
reporting failure on "vectorize_pred"

### PR DESCRIPTION
**Pull request by slomp**
Important bits on lines 71-72 of `vectorize_pred.cpp`; the rest is Visual Studio enforcing some style

### CODE REVIEW

**Comment by shoaibkamil** [comment]
I think the changes are good/necessary, but I have problems with the test-- we really shouldn't be comparing floats for equality, even with predication.

**Comment by steven-johnson** [comment]
ouch, yes, didn't even notice that. exact-float-comparison is a recipe for pain.

**Comment by slomp** [comment]
Should I "epsilonify" the test for now?

**Comment by steven-johnson** [comment]
yes

File: test/performance/vectorize_pred.cpp [hunk 1]
```diff
@@ -62,7 +83,7 @@ bool test(int vec_width) {
 
     for (int y = 0; y < H; y++) {
         for (int x = 0; x < W; x++) {
-            if (outputf(x, y) != outputg(x, y)) {
+            if (!equals(outputf(x, y), outputg(x, y))) {
                 printf("%s x %d failed at %d %d: %d vs %d\n",
```
Comment (shoaibkamil) [inline test/performance/vectorize_pred.cpp]: Given that this test only seems to look at float, perhaps this should be `%f`?

File: test/performance/vectorize_pred.cpp [hunk 2]
```diff
@@ -16,6 +16,27 @@ const char *string_of_type();
...
+}
+
+template<>
+double tolerance<double>() {
+    return 1e-14f;
+}
+
+template<typename T>
+bool equals(T a, T b, T epsilon = tolerance<T>()) {
+    T error = std::abs(a - b);
```
Comment (shoaibkamil) [inline test/performance/vectorize_pred.cpp]: I think this needs specializations for float/double, since you want to call `fabs` there.

**Comment by shoaibkamil** [comment]
@slomp Per conversation with Andrew, can you create an issue for this failing test and move the test to `test/failing_with_issue`?

**Comment by slomp** [comment]
Sure, I'll do that first thing tomorrow morning 👍

File: test/performance/vectorize_pred.cpp (hunk 2 above)
Comment (slomp) [reply test/performance/vectorize_pred.cpp]: `std::abs()` should have the proper overloads for these types.

File: test/performance/vectorize_pred.cpp [hunk 3]
```diff
@@ -1,45 +1,68 @@
...
 #include "halide_benchmark.h"
+#include <cstdio>
 
 using namespace Halide;
 using namespace Halide::Tools;
 
 template<typename A>
 const char *string_of_type();
 
-#define DECL_SOT(name)                                          \
```
Comment (zvookin) [inline test/performance/vectorize_pred.cpp]: This should just go away and the printing should be done using ```Halide::Type``` and the stream ```operator<<``` from IRPrinter.

File: test/performance/vectorize_pred.cpp (hunk 2 above)
Comment (shoaibkamil) [reply test/performance/vectorize_pred.cpp]: You're right, this should be fine.

File: test/performance/vectorize_pred.cpp (hunk 3 above)
Comment (steven-johnson) [reply test/performance/vectorize_pred.cpp]: +1. (I think this is ready to land aside from this nit.)

Comment (slomp) [reply test/performance/vectorize_pred.cpp]: done 👍

**Pull request merged**